src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -0,0 +1,215 @@
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+#nullable enable
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Responsible for orchestrating a source generation pass
+    /// </summary>
+    /// <remarks>
+    /// GeneratorDriver is an immutable class that can be manipulated by returning a mutated copy of itself.
+    /// In the compiler we only ever create a single instance and ignore the mutated copy. The IDE may perform 
+    /// multiple edits, or generation passes of the same driver, re-using the state as needed.
+    /// 
+    /// A generator driver works like a small state machine:
+    ///   - It starts off with no generated sources
+    ///   - A full generation pass will run every generator and produce all possible generated source
+    ///   - At any time an 'edit' maybe supplied, which represents potential future work
+    ///   - TryApplyChanges can be called, which will iterate through the pending edits and try and attempt to 
+    ///     bring the state back to what it would be if a full generation occured by running partial generation
+    ///     on generators that support it
+    ///   - At any time a full generation pass can be re-run, resetting the pending edits
+    /// </remarks>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "In Progress")]
+    public abstract class GeneratorDriver
+    {
+        internal readonly GeneratorDriverState _state;
+
+        internal GeneratorDriver(GeneratorDriverState state)
+        {
+            _state = state;
+        }
+
+        internal GeneratorDriver(Compilation compilation, ParseOptions parseOptions)
+        {
+            _state = new GeneratorDriverState(compilation, parseOptions, ImmutableArray<GeneratorProvider>.Empty, ImmutableArray<AdditionalText>.Empty, ImmutableArray<PendingEdit>.Empty, ImmutableDictionary<GeneratorProvider, ImmutableArray<GeneratedSourceText>>.Empty, finalCompilation: null, editsFailed: true);
+        }
+
+        public GeneratorDriver RunFullGeneration(Compilation compilation, out Compilation outputCompilation, CancellationToken cancellationToken = default)
+        {
+            // with no providers, there is no work to do
+            if (_state.Providers.Length == 0)
+            {
+                outputCompilation = compilation;
+                return this;
+            }
+
+            // PERF: if the input compilation is the same as the last compilation we saw, and we have a final compilation
+            //       we know nothing can have changed and can just short circuit, returning the already processed final compilation 
+            if (compilation == _state.Compilation && _state.FinalCompilation is object)
+            {
+                outputCompilation = _state.FinalCompilation;
+                return this;
+            }
```

Comment:
Do you imagine this optimization is necessary once the IDE is less hacky? We generally hold onto compilations lazily so they can be GCed if not needed in awhile; we might still want to be able to hold onto a GeneratorDriver even if we let the compilation GC since the memory usage might be generally small, relatively.

Reply:
I expect it can go away. We won't need it in the compiler, and when we get it in the IDE properly it should be the case that it never gets hit anyway.

---
In reply to: [388645605](https://github.com/dotnet/roslyn/pull/41789#discussion_r388645605) [](ancestors = 388645605)

Reply:
Alright I'll look to remove this as a part of my work then.